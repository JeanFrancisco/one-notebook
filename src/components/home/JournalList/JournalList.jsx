import { useDispatch, useSelector } from "react-redux";
import JournalEntry from "../JournalEntry/JournalEntry";
import {
    actionReorderByRule,
    actionSetAscendingOrder,
    actionSetDescendingOrder,
} from "../../../redux/actions/notes";

const JournalList = () => {
    const dispatch = useDispatch();
    const items = useSelector( store => store.notes.notes );
    const order = useSelector( store => store.notes.order );
    const orderby = useSelector( store => store.notes.orderby );

    const handleToggleOrderRule = e => {
        dispatch( actionReorderByRule( e.target.value ) );
    }

    const handleToggleOrder = () => {

        if( order === 'asc' ) {
            dispatch( actionSetDescendingOrder() );

            // document.getElementById('btn-sort-icon').setAttribute('class', 'fa fa-sort-amount-up');
            /* Para estos casos. Ambas son malas implementaciones. Ya que se puede usar directamente en la propiedad className. */
            // iconSortRef.current?.setAttribute('class', 'fa fa-sort-amount-up');
        }
        else {
            dispatch( actionSetAscendingOrder() );

            // document.getElementById('btn-sort-icon').setAttribute('class', 'fa fa-sort-amount-down-alt');
            /* Ambas son malas implementaciones. Ya que se puede incluir directamente una condici√≥n en la prop className */
            // iconSortRef.current?.setAttribute('class', 'fa fa-sort-amount-down-alt');
        }
    }

    return <div className="journal__list">

        <div className="journal__list-order-form">
            <label className="journal__list-order-title"> Order by: </label>

            <input
                type="radio"
                name="orderBy"
                value="alphabetic"
                onChange={ handleToggleOrderRule }
                checked={ orderby === 'alphabetic' }
                id="alphabetic"/>
            <label htmlFor="alphabetic" >
                Alphabetically</label>

            <input
                type="radio"
                name="orderBy"
                value="cronologic"
                onChange={ handleToggleOrderRule }
                checked={ orderby === 'cronologic' }
                id="cronologic"/>
            <label htmlFor="cronologic" >
                Cronologically</label>

            <button className="btn" onClick={ handleToggleOrder }>
                <i id="btn-sort-icon"
                    className={ `fa fa-sort-amount-${
                        (order === 'asc')
                        ? 'down-alt'
                        : 'up'
                    }` }>
                </i>
            </button>
        </div>

        {   items.map( note => (
                <JournalEntry key={ note.id } entry={ note } />
            ))
        }
    </div>
}

export default JournalList;