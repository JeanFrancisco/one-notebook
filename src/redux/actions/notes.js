import Swal from 'sweetalert2';
import { addDoc, collection, deleteDoc, doc, getDocs, orderBy, updateDoc } from '@firebase/firestore';
import {
    NOTES_NEW_NOTE,
    NOTES_SET_ACTIVE,
    NOTES_DISPLAY_IN_SIDEBAR,
    NOTES_UPDATED_NOTE,
    NOTES_DROP_NOTE,
    NOTES_CLEANING_WHEN_LOGOUT,
    NOTES_CHANGE_ORDER,
} from '../constants/typesNotes';
import { uploadFile } from '../../helpers/files';
import { db as firestoreDB } from '../../config/firebase';

export const actionAddNewNote = () => {
    return async (dispatch, getState) => {
        const { uid: user_uid } = getState().auth;

        const newNote = {
            title: '',
            body: '',
            urlImage: '',
            createdAt: new Date().getTime(),
        };

        const doc = await addDoc( collection(firestoreDB, `${user_uid}/journal/notes`), newNote);

        const createdNote = {
            id: doc.id,
            ...newNote,
        };

        dispatch( actionNewNote( createdNote ) );
        dispatch( actionSetActiveNote( createdNote ) );
    }
};

export const actionNewNote = ( createdNote ) => ({
    type: NOTES_NEW_NOTE,
    payload: createdNote,
});

export const actionSetActiveNote = ( activeNote ) => ({
    type: NOTES_SET_ACTIVE,
    payload: activeNote,
});

export const actionSavingChanges = () => {
    return async (dispatch, getState) => {
        const user_uid = getState().auth.uid;
        const activeNote = getState().notes.active;

        const { id: note_id, ...noteForFirestore } = activeNote;
        await updateDoc( doc(firestoreDB, `${user_uid}/journal/notes/${ note_id }`), noteForFirestore )
            .then(() => {
                dispatch( refreshNote(activeNote) );

                Swal.fire({
                    title: 'Guardada exitosamente',
                    html: '<p><small>Su nota se guardó correctamente</small></p>',
                    icon: 'success',
                    timer: 1300,
                    showConfirmButton: true,
                    confirmButtonText: 'Aceptar',
                    confirmButtonColor: 'limegreen',
                });
            })
            .catch(reason => {
                console.log( reason.code );
            });
    }
}

export const refreshNote = ( updatedNote ) => ({
    type: NOTES_UPDATED_NOTE,
    payload: { ...updatedNote },
});

export const getFirebaseNotes = () => {
    return async ( dispatch, getState ) => {
        const { uid } = getState().auth;

        const snapNotes = await getDocs( collection(firestoreDB, `${uid}/journal/notes`) );
        const notes = [];

        snapNotes.forEach( snapHijo => {
            notes.push({
                id: snapHijo.id,
                ...snapHijo.data(),
            })
        });

        dispatch( displayNotes(notes) );
    }
}

export const displayNotes = ( notes ) => ({
    type: NOTES_DISPLAY_IN_SIDEBAR,
    payload: notes,
});

export const actionSetAscendingOrder = () => ({
    type: NOTES_CHANGE_ORDER,
    payload: {
        order: 'asc',
    }
});

export const actionSetDescendingOrder = () => ({
    type: NOTES_CHANGE_ORDER,
    payload: {
        order: 'desc',
    }
});

export const actionReorderByRule = ( orderBy ) => ({
    type: NOTES_CHANGE_ORDER,
    payload: {
        orderby: orderBy,
    }
});

export const actionStartUploadFile = (file) => {
    return async (dispatch, getState) => {
 
        Swal.fire({
            title: 'Subiendo archivos...',
            text: 'Por favor, espere un momento...',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () => {
                Swal.showLoading();
            },
        });

        const activeNote = getState().notes.active;
        const { secure_url } = await uploadFile(file);

        dispatch( actionSetActiveNote({ ...activeNote, urlImage: secure_url }) );
        await dispatch( actionSavingChanges() );

        Swal.close();
    }
}

export const actionStartDeletingNote = (note_id) => {
    return async (dispatch, getState) => {
        const user_uid = getState().auth.uid;

        try {
            const resp = await deleteDoc( doc( firestoreDB, `${user_uid}/journal/notes/${note_id}`) );

            dispatch( deleteNote( note_id ) );

            Swal.fire({
                icon: 'success',
                timer: 1250,
                title: 'Eliminado',
                text: 'Su nota fue eliminada',
                confirmButtonText: 'Aceptar',
                confirmButtonColor: 'green',
                showCloseButton: true,
            });

        } catch (e) {
            Swal.fire({
                icon: 'error',
                html: 'Error en la conexión con Firestore DB',
                title: 'Error. <small>No pudimos terminar la eliminación</small>',
                footer: `<small>For more details: ${e.code}</small>`,
                confirmButtonText: 'Entendido',
                confirmButtonColor: 'oragered', 
                showCloseButton: true,
            });

            console.error(e);
        }
    }
}

export const deleteNote = (noteID) => ({
    type: NOTES_DROP_NOTE,
    payload: noteID
});

export const cleanupNotesLogout = () => ({
    type: NOTES_CLEANING_WHEN_LOGOUT
});