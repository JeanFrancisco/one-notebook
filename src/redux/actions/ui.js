import { UI_FINISH_LOADING, UI_START_LOADING } from "../constants/typesUi";

export const actionStartLoading = () => ({
    type: UI_START_LOADING,
});

export const actionFinishLoading = () => ({
    type: UI_FINISH_LOADING,
});