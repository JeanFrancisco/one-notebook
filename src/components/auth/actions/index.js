import {
    createUserWithEmailAndPassword,
    signInWithEmailAndPassword,
    GoogleAuthProvider,
    signInWithPopup,
    updateProfile,
    getAuth,
    signOut,
} from "firebase/auth";
import Swal from 'sweetalert2';
import { app } from "../../../config/firebase";
import { cleanupNotesLogout } from "../../../redux/actions/notes";
import { actionFinishLoading, actionStartLoading } from "../../../redux/actions/ui";
import {
    AUTH_LOGIN,
    AUTH_LOGOUT,
    AUTH_SET_ERROR,
    AUTH_REMOVE_ERROR
} from "../constants";

export const authLogin = (uid, displayName) => ({
    type: AUTH_LOGIN,
    payload: {
        uid,
        displayName,
    }
});

export const authSignInWithGoogle = () => {
    return async (dispatch) => {

        dispatch( actionStartLoading() );

        const auth = getAuth( app );

        await signInWithPopup( auth, new GoogleAuthProvider() )
        .then( userCredentials => {
            // const { user: { displayName, uid } } = userCredentials;
            
            // dispatch( authLogin(uid, displayName) );
        })
        .catch( reason => {
            Swal.fire({
                icon: 'warning',
                html: '<small>No se pudo iniciar la sesión. <br/> <br/> ¿Alcanzó a ver si ocurrió algo: con el Pop-up de autenticación de Google?</small>',
                title: 'Oups! Lo sentimos.',
                iconHtml: '<i class="fa fa-warning"></i>',
                footer: `<small>Reason Code: ${reason.code}</small>`,
                showCloseButton: true,
                confirmButtonText: 'Aceptar',
            });
        })
        .finally(() => {
            dispatch( actionFinishLoading() );
        });
    };
}

export const authEmailPasswordSignIn = ({ email, password }) => {
    return async dispatch => {
        dispatch( actionStartLoading() );

        const auth = getAuth(app);

        await signInWithEmailAndPassword(auth, email, password)
            .then( ({ user }) => {
                // const { uid, displayName } = user;

                // dispatch( authLogin(uid, displayName) );
            }).catch( reason => {
                Swal.fire({
                    title: 'Oups! <small>Algo ocurrió mientras iniciabamos tu sesión.</small>',
                    html: `<p><b><small>¿Qué acciones puedes hacer?</small></b></p>
                        <ul style="text-align: left">
                            <li><small>Puedes revisar por favor tu email y contraseña.</small></li>
                            <li><small>También que te puedes conectar a Internet. En especial a los servicios de Google.</small></li>
                        </ul>`,
                    confirmButtonText: 'Entendido',
                    icon: 'warning',
                    iconColor: 'orange',
                    showCloseButton: true,
                    confirmButtonColor: 'orangered',
                    footer: `<small>Reason Code: ${reason.code}</small>`,
                })
            }).finally(() => {
                dispatch( actionFinishLoading() );
            });
    }
}

export const authEmailPasswordSignUp = ({ email, password, username }) => {
    return async (dispatch) => {
        dispatch( actionStartLoading() );

        const auth = getAuth(app);

        await createUserWithEmailAndPassword(auth, email, password)
            .then( async userCredentials => {

                const { user } = userCredentials;
                await updateProfile(user, { displayName: username });

                // dispatch( authLogin(user.uid, user.displayName) );

            }).catch( reason => {
                Swal.fire({
                    title: 'Oups! <small>Ahora mismo no podemos registrarte.</small>',
                    html: '<p>Por favor... <br/> <small>Regresa más tarde y vuelve a intentarlo ;) .</small><p>',
                    icon: 'info',
                    iconColor: 'orange',
                    confirmButtonText: 'Entiendo',
                    confirmButtonColor: 'limegreen',
                    showCloseButton: true,
                    footer: `<small>Reason Code: ${reason.code}</small>`,
                });
            }).finally(() => {
               dispatch( actionFinishLoading() );
            });
    }
}

export const authFirebaseLogout = () => {
    return async dispatch => {
        const auth = getAuth(app);

        await signOut(auth);

        dispatch( authLogout() );

        dispatch( cleanupNotesLogout() );
    }
}

export const authLogout = () => ({ type: AUTH_LOGOUT });

export const setAuthError = ( msg ) => ({
    type: AUTH_SET_ERROR,
    payload: msg
});

export const removeAuthError = () => ({ type: AUTH_REMOVE_ERROR });