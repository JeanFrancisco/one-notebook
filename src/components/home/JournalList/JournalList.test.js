import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import JournalList from './JournalList';
import { actionReorderByRule, actionSetDescendingOrder, actionSetAscendingOrder } from '../../../redux/actions/notes';

const mockStore = configureStore([ thunk ]);

const store = mockStore({
    notes: {
        notes: [],
        order: 'asc',
        orderby: 'cronologic'
    }
});

store.dispatch = jest.fn();

jest.mock('../../../redux/actions/notes', () => ({
    actionReorderByRule: jest.fn(),
    actionSetDescendingOrder: jest.fn(),
    actionSetAscendingOrder: jest.fn(),
}));

describe('Pruebas al componente de <JournalList/>.jsx', () => {

    const wrapper = mount(<Provider store={ store }>
        <JournalList />
    </Provider>);

    beforeEach(() => {
        jest.clearAllMocks();
    })
    
    test('Debe hacer match con el snapshot por defecto.', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe hacer match con el snapshot por defecto. Nótese que el order se establece a desc y el orderby a alphabetic', () => {
        const notes = [
            { id: '1', title: 'Test title One', body: 'Lorem impsu' },
            { id: '2', title: 'Test title Two', body: 'Lorem impsu' },
            { id: '3', title: 'Test title Three', body: 'Lorem impsu' },
            { id: '4', title: 'Test title Four', body: 'Lorem impsu' },
        ];

        const store = mockStore({
            notes: {
                notes: notes,
                order: 'desc',
                orderby: 'alphabetic'
            }
        });

        const wrapper = mount(<Provider store={ store }>
            <JournalList />
        </Provider>);

        expect( wrapper ).toMatchSnapshot();
    });


    test('Debe renderizar tantos componentes <JournalEntry/>, como número de elementos haya en el arreglo de notes', () => {
        const notes = [
            { id: '1', title: 'Test title One', body: 'Lorem impsu' },
            { id: '2', title: 'Test title Two', body: 'Lorem impsu' },
            { id: '3', title: 'Test title Three', body: 'Lorem impsu' },
            { id: '4', title: 'Test title Four', body: 'Lorem impsu' },
        ];

        const store = mockStore({
            notes: {
                notes: notes,
                order: 'asc',
                orderby: 'cronologic'
            }
        });

        const wrapper = mount(<Provider store={ store }>
            <JournalList />
        </Provider>);

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('JournalEntry') ).toHaveLength( notes.length );
    });

    test(`Debe llamar a la acción para cambiar el criterio de ordenamiento, pasándole dicho criterio como parámetro.
    Al cambiar el valor del input radio (onChange event)`, () => {

        const criteria = 'alphabetic';

        wrapper.find('#alphabetic').prop('onChange')({
            target: {
                value: criteria,
            }
        });

        expect( actionReorderByRule ).toHaveBeenCalledTimes(1);
        expect( actionReorderByRule ).toHaveBeenCalledWith( criteria )
    });

    test(`Debe llamar a las acciones de cambiar el orden, ascendente o descendente. Al dar click en el botón.`, () => {

        const wrapper = mount(<Provider store={ store }>
            <JournalList />
        </Provider>);

        // wrapper.find('#btn-sort-icon').closest('button').prop('onClick')();
        wrapper.find('#btn-sort-icon').closest('button').simulate('click');

        expect( actionSetDescendingOrder ).toHaveBeenCalledTimes(1);

        const new_store = mockStore({
            notes: {
                notes: [],
                order: 'desc',
                orderby: 'alphabetic',
            }
        });

        new_store.dispatch = jest.fn();

        const new_wrapper = mount(<Provider store={ new_store }>
            <JournalList />
        </Provider>);

        new_wrapper.find('#btn-sort-icon').parent().prop('onClick')();

        expect( actionSetAscendingOrder ).toHaveBeenCalledTimes(1);
    });
});