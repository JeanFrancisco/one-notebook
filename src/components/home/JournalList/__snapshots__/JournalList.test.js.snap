// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas al componente de <JournalList/>.jsx Debe hacer match con el snapshot por defecto. 1`] = `
<div
  className="journal__list"
>
  <div
    className="journal__list-order-form"
  >
    <label
      className="journal__list-order-title"
    >
       Order by: 
    </label>
    <input
      checked={false}
      id="alphabetic"
      name="orderBy"
      onChange={[Function]}
      type="radio"
      value="alphabetic"
    />
    <label
      htmlFor="alphabetic"
    >
      Alphabetically
    </label>
    <input
      checked={true}
      id="cronologic"
      name="orderBy"
      onChange={[Function]}
      type="radio"
      value="cronologic"
    />
    <label
      htmlFor="cronologic"
    >
      Cronologically
    </label>
    <button
      className="btn"
      onClick={[Function]}
    >
      <i
        className="fa fa-sort-amount-down-alt"
        id="btn-sort-icon"
      />
    </button>
  </div>
</div>
`;

exports[`Pruebas al componente de <JournalList/>.jsx Debe hacer match con el snapshot por defecto. Nótese que el order se establece a desc y el orderby a alphabetic 1`] = `
<div
  className="journal__list"
>
  <div
    className="journal__list-order-form"
  >
    <label
      className="journal__list-order-title"
    >
       Order by: 
    </label>
    <input
      checked={true}
      id="alphabetic"
      name="orderBy"
      onChange={[Function]}
      type="radio"
      value="alphabetic"
    />
    <label
      htmlFor="alphabetic"
    >
      Alphabetically
    </label>
    <input
      checked={false}
      id="cronologic"
      name="orderBy"
      onChange={[Function]}
      type="radio"
      value="cronologic"
    />
    <label
      htmlFor="cronologic"
    >
      Cronologically
    </label>
    <button
      className="btn"
      onClick={[Function]}
    >
      <i
        className="fa fa-sort-amount-up"
        id="btn-sort-icon"
      />
    </button>
  </div>
  <div
    className="journal__entry animate__animated animate__pulse"
    onClick={[Function]}
  >
    <div
      className="journal__entry-content"
    >
      <h5
        className="journal__entry-title"
      >
        Test title One
      </h5>
      <p
        className="journal__entry-description"
      >
        Lorem impsu
      </p>
    </div>
    <label
      className="journal__entry-date"
    >
      sábado
      <span
        className="journal__entry-day"
      >
        5
      </span>
    </label>
  </div>
  <div
    className="journal__entry animate__animated animate__pulse"
    onClick={[Function]}
  >
    <div
      className="journal__entry-content"
    >
      <h5
        className="journal__entry-title"
      >
        Test title Two
      </h5>
      <p
        className="journal__entry-description"
      >
        Lorem impsu
      </p>
    </div>
    <label
      className="journal__entry-date"
    >
      sábado
      <span
        className="journal__entry-day"
      >
        5
      </span>
    </label>
  </div>
  <div
    className="journal__entry animate__animated animate__pulse"
    onClick={[Function]}
  >
    <div
      className="journal__entry-content"
    >
      <h5
        className="journal__entry-title"
      >
        Test title Three
      </h5>
      <p
        className="journal__entry-description"
      >
        Lorem impsu
      </p>
    </div>
    <label
      className="journal__entry-date"
    >
      sábado
      <span
        className="journal__entry-day"
      >
        5
      </span>
    </label>
  </div>
  <div
    className="journal__entry animate__animated animate__pulse"
    onClick={[Function]}
  >
    <div
      className="journal__entry-content"
    >
      <h5
        className="journal__entry-title"
      >
        Test title Four
      </h5>
      <p
        className="journal__entry-description"
      >
        Lorem impsu
      </p>
    </div>
    <label
      className="journal__entry-date"
    >
      sábado
      <span
        className="journal__entry-day"
      >
        5
      </span>
    </label>
  </div>
</div>
`;

exports[`Pruebas al componente de <JournalList/>.jsx Debe renderizar tantos componentes <JournalEntry/>, como número de elementos haya en el arreglo de notes 1`] = `
<div
  className="journal__list"
>
  <div
    className="journal__list-order-form"
  >
    <label
      className="journal__list-order-title"
    >
       Order by: 
    </label>
    <input
      checked={false}
      id="alphabetic"
      name="orderBy"
      onChange={[Function]}
      type="radio"
      value="alphabetic"
    />
    <label
      htmlFor="alphabetic"
    >
      Alphabetically
    </label>
    <input
      checked={true}
      id="cronologic"
      name="orderBy"
      onChange={[Function]}
      type="radio"
      value="cronologic"
    />
    <label
      htmlFor="cronologic"
    >
      Cronologically
    </label>
    <button
      className="btn"
      onClick={[Function]}
    >
      <i
        className="fa fa-sort-amount-down-alt"
        id="btn-sort-icon"
      />
    </button>
  </div>
  <div
    className="journal__entry animate__animated animate__pulse"
    onClick={[Function]}
  >
    <div
      className="journal__entry-content"
    >
      <h5
        className="journal__entry-title"
      >
        Test title One
      </h5>
      <p
        className="journal__entry-description"
      >
        Lorem impsu
      </p>
    </div>
    <label
      className="journal__entry-date"
    >
      sábado
      <span
        className="journal__entry-day"
      >
        5
      </span>
    </label>
  </div>
  <div
    className="journal__entry animate__animated animate__pulse"
    onClick={[Function]}
  >
    <div
      className="journal__entry-content"
    >
      <h5
        className="journal__entry-title"
      >
        Test title Two
      </h5>
      <p
        className="journal__entry-description"
      >
        Lorem impsu
      </p>
    </div>
    <label
      className="journal__entry-date"
    >
      sábado
      <span
        className="journal__entry-day"
      >
        5
      </span>
    </label>
  </div>
  <div
    className="journal__entry animate__animated animate__pulse"
    onClick={[Function]}
  >
    <div
      className="journal__entry-content"
    >
      <h5
        className="journal__entry-title"
      >
        Test title Three
      </h5>
      <p
        className="journal__entry-description"
      >
        Lorem impsu
      </p>
    </div>
    <label
      className="journal__entry-date"
    >
      sábado
      <span
        className="journal__entry-day"
      >
        5
      </span>
    </label>
  </div>
  <div
    className="journal__entry animate__animated animate__pulse"
    onClick={[Function]}
  >
    <div
      className="journal__entry-content"
    >
      <h5
        className="journal__entry-title"
      >
        Test title Four
      </h5>
      <p
        className="journal__entry-description"
      >
        Lorem impsu
      </p>
    </div>
    <label
      className="journal__entry-date"
    >
      sábado
      <span
        className="journal__entry-day"
      >
        5
      </span>
    </label>
  </div>
</div>
`;
